### ISSUE TITLE
Callable is never an instance of class

### ISSUE THREAD

**Issue by Timmmm** [microsoft/pyright#10171]
```python
from typing import Any

class Foo:
    pass

def f(x: Any):
    if callable(x):
        if isinstance(x, Foo):
            print("Ok")

class Bar(Foo):
    def __call__(self, *args: Any, **kwds: Any) -> Any:
        print("Called")

b = Bar()

f(b)
```

This gives me a false positive here:

![Image](https://github.com/user-attachments/assets/849c3730-1de3-46e2-a258-fcf75f6f86a0)

```
Unnecessary isinstance call; "(...) -> object" is never an instance of "Foo"
```

If I run it it works fine (prints `Ok` with no errors).

The Python docs [say `callable()` means the value has a `__call__()` method](https://docs.python.org/3/library/functions.html#callable), which can totally be the case for classes.

If I change `isinstance(x, Foo)` to `isinstance(x, Bar)` then the error goes away, so I think it's forgetting that `isinstance(x, Foo)` means `Foo or derived from Foo`. If it was `if type(x) == Foo` then the error would be correct.

Using VSCode Python extension version 2025.2.0

**Comment by erictraut** [comment]
I agree this is a bug. The type shouldn't be narrowed to `Never` in this case. Ideally, it would be narrowed to a type that is an intersection between a callable (`(...) -> Any`) and `Foo`. Since we don't have support for intersections in the type system currently, we need to pick one of the two subtypes. Mypy appears to narrow to `Foo` in this case. I tried it both ways, and mypy's method seems to produce the fewest false positives, so I went with that approach.

One exception is if the class is `@final`, then it should still be narrowed to `Never` since there's no way for a subclass to supply a `__call__` method in this case.

**Comment by Timmmm** [comment]
Thanks so much! Impossibly speedy as ever. I hope they're paying you 10x salary to go with that 10x engineering! :-D

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.399.
